cachito/web/content_manifest.py

```diff
@@ -232,3 +237,246 @@ def generate_icm(self, image_contents=None):
         icm["image_contents"] = image_contents or []
         return deep_sort_icm(icm)
+
+
+class ContentManifestPackage:
+    """
+    A package within a content manifest.
+
+    It is used primarily to generate a package URL (purl).
+    """
+
+    def __init__(self, name, type, version, dev=False, dependencies=[], path=None, replaces=None):
+        """
+        Initialize ContentManifestPackage.
+
+        :param str name: the package's name
+        :param str type: the package manager used to process this package. Should be one
+            of the following: gomod, go-package, npm, yarn, pip or git-submodule
+        :param str version: the package's version
+        :param bool dev: if the package is only used in a dev environment
+        :param list[dict] dependencies: the package's dependencies as a list of JSON
+            dicts
+        :param str path: relative path to package from root of repository
+        :param str replaces: the dependency that is being replaced by this one in the
+            context of the request
+        """
+        self.name = name
+        self.type = type
+        self.version = version
+        self.dev = dev
+        self.path = path
+        self.replaces = replaces
+
+        self.dependencies = [
+            ContentManifestPackage.from_json(dependency) for dependency in dependencies
+        ]
+
+    def __repr__(self):
+        return (
+            f"<{self.__class__.__name__} name={self.name}, type={self.type}, "
+            f"version={self.version}>"
+        )
+
+    @property
+    def hash(self):
+        """Generate the package hash based on its name, type and version."""
+        return hash(f"{self.name}{self.type}{self.version}")
+
+    @classmethod
+    def validate_json(cls, package):
+        """
+        Validate the JSON representation of a package.
+
+        :param dict package: the JSON representation of a package
+        :raise ValidationError: if the JSON does not match the required schema
+        """
+        required = ("name", "type", "version")
+
+        if not isinstance(package, dict):
+            raise ValidationError("A package JSON must be an instance of dict")
+
+        for key in required:
+            if key not in package:
+                raise ValidationError(
+                    "A package must be a JSON object with the following "
+                    f'keys: {", ".join(sorted(required))}.'
+                )
+            if not isinstance(package[key], str):
+                raise ValidationError('The "{}" key of the package must be a string'.format(key))
+
+    @classmethod
+    def from_json(cls, package):
+        """
+        Create a Package object from JSON.
+
+        :param dict package: the dictionary representing the package
+        :return: the Package object
+        :rtype: ContentManifestPackage
+        """
+        cls.validate_json(package)
+        return cls(**package)
+
+    def to_json(self, dependencies=None, subpath=None):
```

Comment:
Is this ever used?

Reply:
Apparently, no :grin:

Reply:
In fact, it was being used in a test :grin: 

I don't think it was enough to justify the function existence, though.